src/regex.cpp

```diff
@@ -191,7 +191,9 @@ BOOST_REGEX_DECL void BOOST_REGEX_CALL put_mem_block(void* p)
 
 #else
 
-#ifdef BOOST_HAS_THREADS
+#if defined(BOOST_REGEX_MEM_BLOCK_CACHE_LOCK_FREE)
+mem_block_cache block_cache;
```

Comment:
Hmm.... That's messy. A possibility is this:

Change the constructor of mem_block_cache to a member function, say... initialize()

Add a std::once_flag to the mem_block_cache object.

In the get() and put() function use std::call_once to call initialize()

Assuming std::call_once / boost::call_once has a well implemented fast-path, I think this should work and maintain performance.